internal/js/listener_js.go

```diff
@@ -0,0 +1,100 @@
+package js
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"reflect"
+	"syscall/js"
+)
+
+const (
+	// jsFieldCallbackID is field name of js.Func struct that contains callback ID of Go func exposed to JS.
+	jsFieldCallbackID = "id"
+
+	// msgBufferSize is size of a buffer of incoming messages from LSP client.
+	msgBufferSize = 10
+)
+
+// registerRequestListener creates and registers LSP request listener at the host.
+// This method should only be called once at start of a server.
+//
+// Context parameter is used to automatically dispose underlying channel and callback.
+//
+// Returns a reader to consume incoming JSON-RPC messages from host.
+// Closing a returned reader acts the same as cancelling an input context.
+func registerRequestListener(ctx context.Context) (io.ReadCloser, error) {
+	chanCtx, cancelFn := context.WithCancel(ctx)
+
+	inputEvents := make(chan []byte, msgBufferSize)
+	callback := js.FuncOf(func(this js.Value, args []js.Value) any {
+		if len(args) == 0 {
+			// this is the only way to throw a JS exception.
+			panic("missing argument")
+		}
+		message := args[0].String()
+		inputEvents <- []byte(message)
+		return nil
+	})
+
+	// As it's impossible to pass JS objects to import funcs.
+	// Obtain the function callback ID pass it to import.
+	callbackID, err := getFuncCallbackID(callback)
+	if err != nil {
+		cancelFn()
+		callback.Release()
+		close(inputEvents)
+		return nil, fmt.Errorf("failed to prepare callback: %w", err)
+	}
+
+	registerCallback(callbackID)
+
+	go func() {
+		select {
+		case <-chanCtx.Done():
+			callback.Release()
+			close(inputEvents)
+		}
```

Comment:
Could be simplified to:

```suggestion
		<-chanCtx.Done()
		callback.Release()
		close(inputEvents)
```